pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,286 @@
+	operators.ImageOperatorInstance, error,
+) {
+	instance := &wasmOperatorInstance{
+		gadgetCtx: gadgetCtx,
+		handleMap: map[uint32]any{},
+		logger:    gadgetCtx.Logger(),
+	}
+
+	if err := instance.init(gadgetCtx, target, desc); err != nil {
+		instance.close(gadgetCtx)
+		return nil, fmt.Errorf("initializing wasm: %w", err)
+	}
+
+	return instance, nil
+}
+
+type wasmOperatorInstance struct {
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+
+	dsCallback wapi.Function
+
+	// Golang objects are exposed to the wasm module by using a handleID
+	handleMap       map[uint32]any
+	lastHandleIndex uint32
+	handleLock      sync.RWMutex
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	if err := i.callGuestFunction(gadgetCtx.Context(), "init"); err != nil {
+		return fmt.Errorf("initializing wasm guest: %w", err)
+	}
+
+	return nil
+}
+
+func (i *wasmOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (i *wasmOperatorInstance) addHandle(obj any) uint32 {
+	if obj == nil {
+		return 0
+	}
+
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+
+	if len(i.handleMap) == maxHandles {
+		i.logger.Warnf("too many open handles")
+		return 0
+	}
+
+	handleIndex := i.lastHandleIndex
+	handleIndex++
+
+	// look for a free index in the map
+	for {
+		// zero is reserved, handle overflow
+		if handleIndex == 0 {
+			handleIndex++
+		}
+
+		if _, ok := i.handleMap[handleIndex]; !ok {
+			// register new entry
+			i.handleMap[handleIndex] = obj
+			i.lastHandleIndex = handleIndex
+			return handleIndex
+		}
+		handleIndex++
+	}
+}
+
+// getHandleTyped returns the handle with the given ID, casted to the given type.
+// It can't be implemented as a generic method because it's not supported by Go yet.
+func getHandle[T any](i *wasmOperatorInstance, handleID uint32) (T, bool) {
+	i.handleLock.RLock()
+	defer i.handleLock.RUnlock()
+
+	var empty T // zero value
+
+	val, ok := i.handleMap[handleID]
+	if !ok {
+		i.logger.Warnf("handle %d not found", handleID)
+		return empty, false
+	}
+
+	t, ok := val.(T)
+	if !ok {
+		i.logger.Warnf("bad handle type for %d", handleID)
+		return empty, false
+	}
+
+	return t, true
+}
+
+func (i *wasmOperatorInstance) delHandle(handleID uint32) {
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	delete(i.handleMap, handleID)
+}
+
+func (i *wasmOperatorInstance) init(
+	gadgetCtx operators.GadgetContext,
+	target oras.ReadOnlyTarget,
+	desc ocispec.Descriptor,
+) error {
+	ctx := gadgetCtx.Context()
+	rtConfig := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(256) // 16MB (64KB per page)
+	i.rt = wazero.NewRuntimeWithConfig(ctx, rtConfig)
+
+	env := i.rt.NewHostModuleBuilder("env")
+
+	i.addLogFuncs(env)
```

Comment:
Maybe we should have one more obligatory function that wasm has to export: `func getVersion() int`. We could call it here and make sure it returns the version (or _A_ version) that the WASM operator can deal with. Whenever we introduce breaking changes to the API, we could (probably in most cases) still support older version for a certain period.

Reply:
I added it. It's very strict for now (requires an exact match on an integer), but we can make it more flexible if needed later on.